### PR TITLE
CNI: Move ztunnel redirection config out of init container and into CNI plugin 

### LINKED ISSUE #41917: Support AKS for ambient
### Bug Description

Installing Istio Ambient in AKS, as soon as the namespace is label istio.io/dataplane-mode=ambient the Pods lose all the connectivity.

I tracked down the problem to the geneve tunnel between the Kubernetes node and the ztunnel pod to be broken.

Doing a `tcpdump` in the `ztunnel` pod on interface `eth0` I see geneve traffic coming in, where the source IP address is the IP address of the Kubernetes node. In this example 10.240.0.4 is the IP address of the kubernetes node and 10.240.0.52 is the IP address of the `ztunnel` pod:

```
root@ztunnel-f4d6p:/# tcpdump -i eth0 -n udp
tcpdump: verbose output suppressed, use -v[v]... for full protocol decode
listening on eth0, link-type EN10MB (Ethernet), snapshot length 262144 bytes
21:01:14.807437 IP 10.240.0.4.61534 > 10.240.0.52.6081: Geneve, Flags [none], vni 0x3e9: ARP, Request who-has 192.168.127.2 tell 192.168.127.1, length 28
21:01:15.834678 IP 10.240.0.4.61534 > 10.240.0.52.6081: Geneve, Flags [none], vni 0x3e9: ARP, Request who-has 192.168.127.2 tell 192.168.127.1, length 28
```

The traffic is not passed to the geneve interface `istioin` because the geneve tunnel is setup incorrectly:

The `ztunnel` initContainer has this code:

https://github.com/istio/istio/blob/23bfa062c7107a45e6b863c9f976cf1fd22f7e26/manifests/charts/ztunnel/templates/daemonset.yaml#L116-L125

There is a problem with the line:
```
HOST_IP=$(ip route | grep default | awk '{print $3}')
```

In Microsoft Azure AKS `HOST_IP` will always get the value 169.254.1.1

Example routing table inside the `ztunnel` pod:

```
root@ztunnel-f4d6p:/# ip route
default via 169.254.1.1 dev eth0 proto static
169.254.1.1 dev eth0 proto static scope link
192.168.126.0/30 dev pistioin proto kernel scope link src 192.168.126.2
192.168.127.0/30 dev pistioout proto kernel scope link src 192.168.127.2
```

The script should be portable to detect the `HOST_IP` correctly on different Kubernetes platforms

### Version

```prose
go run ./istioctl/cmd/istioctl version
client version: unknown
control plane version: 1.17
data plane version: 1.17-dev (3 proxies)

kubectl version --short
Flag --short has been deprecated, and will be removed in the future. The --short output will become the default.
Client Version: v1.25.3
Kustomize Version: v4.5.7
Server Version: v1.25.2
```

### Additional Information

_No response_

### PR DESCRIPTION
**Pull request by bleggett**
This implements https://github.com/istio/istio/issues/43411, among other things.

Previously, the CNI plugin would configure ambient networking in the node netns for application pods and `ztunnel`, and in the pod netns for application pods. `ztunnel`'s networking config was still using an init container to configure iptables/etc.

This moves the netconfig (for both eBPF and iptables) that existed in the `ztunnel` init container into the CNI plugin and removes the init container - the CNI plugin drops into the ztunnel netns and configures it as needed.

- Removes the need for a `ztunnel` init container and scattering interdependent hardcoded rules across CNI plugin and helm charts.
- This means we should be OK to use `distroless` for the ztunnel base image now (don't need `ip` or `iptables` in the ztunnel image anymore)
- This also means the `ztunnel` daemonset no longer needs elevated privileges to mutate its own networking rules.
- Uses the more reliable HOST_IP detection method introduced in https://github.com/istio/istio/pull/42372, which should fix https://github.com/istio/istio/issues/41917 and generally provide a consistent way to get HOST_IP regardless of platform, as well (need to do a manual test on AKS to prove it).

Additional changes:

- The original goal (I think) was to 
  - keep "generic" net config funcs in `net.go`
  - Linux-specific ones in `net_linux.go`
  - potentially support other OS networking stacks with `net_xx.go` parallel files. 
- The actual evolution has resulted in `net.go` being entirely and unavoidably Linux specific, with no clean logical separation between it and `net_linux.go`. I have attempted to move most of the obviously Linux-specific helper funcs still left in `net.go` into `net_linux.go`, but I am not convinced the separation amounts to much.
-   The entire CNI plugin is unarguably Linux-specific at this time, so I have deleted `net_unspecified.go` - practically it accomplished little and is a poor substitute for a formal interface definition. A better approach for supporting alternate OSes would be an entirely separate purpose-built CNI plugin IMO, as virtually nothing from this CNI plugin would be practically reusable for a non-Linux OS.

**To help us figure out who should review this PR, please put an X in all the areas that this PR affects.**

- [x] Ambient
- [ ] Configuration Infrastructure
- [ ] Docs
- [ ] Installation
- [x] Networking
- [ ] Performance and Scalability
- [ ] Policies and Telemetry
- [ ] Security
- [ ] Test and Release
- [ ] User Experience
- [ ] Developer Infrastructure

**Please check any characteristics that apply to this pull request.**

- [x] Does not have any [user-facing](https://github.com/istio/istio/tree/master/releasenotes#when-to-add-release-notes) changes. This may include CLI changes, API changes, behavior changes, performance improvements, etc.

### CODE REVIEW

**Comment by bleggett** [comment]
@irisdingbj and @PlatformLC - please take a look when you have a chance, as this makes use of some of the helper code you added for eBPF.

**Comment by PlatformLC** [comment]
> @irisdingbj and @PlatformLC - please take a look when you have a chance, as this makes use of some of the helper code you added for eBPF.

LGTM for moving the rules config from init container to CNI

**Comment by costinm** [comment]
I agree, other platforms can use a different CNI pod image. 

Question: would it make sense ( not in 1.18, later) to cleanup a bit the cni plugin, and have a set of smaller specialized ones ? I liked the script's readability - mixing redirect, tproxy, ebpf in same code may become confusing... 
We have discussed adding a CRD to control interception - exclude ports in particular for proxyless apps ( also after we get initial ambient out ).

I will file a bug/feature request for handling dynamic redirection changes for the case ztunnel is uograded - not sure if this was decided, but I suspect using cni may be better/simer than a hot restart of ztunnel.

**Comment by bleggett** [comment]
> I agree, other platforms can use a different CNI pod image.
> 
> Question: would it make sense ( not in 1.18, later) to cleanup a bit the cni plugin, and have a set of smaller specialized ones ? I liked the script's readability - mixing redirect, tproxy, ebpf in same code may become confusing... We have discussed adding a CRD to control interception - exclude ports in particular for proxyless apps ( also after we get initial ambient out ).

Smaller specialized separate ones might be worth looking into - additionally, we might actually consider dropping/deprecating certain CNI mechanisms (non-eBPF ambient, ambient and sidecar deduplicating a bit) eventually, which would also help.

Anything touching the OS networking stack will always have some basic level of complexity to your point, and it's worth our time to control for that and make it easier to reason about and work on.

I don't have a good understanding of how a CRD would practically simplify things versus the existing daemonset - most of the complexity is in manipulating the networking stack and that would be present in a CRD as well - but if we think it can simplify things it's worth thinking about.

> I will file a bug/feature request for handling dynamic redirection changes for the case ztunnel is uograded - not sure if this was decided, but I suspect using cni may be better/simer than a hot restart of ztunnel.

Yes.

File: manifests/charts/istio-cni/templates/daemonset.yaml [hunk 1]
```diff
@@ -170,44 +170,6 @@ spec:
 {{- else }}
 {{ toYaml .Values.global.defaultResources | trim | indent 12 }}
 {{- end }}
-          {{- if .Values.cni.ambient.enabled }}
```
[the remark this answers is not in the document]
Comment (howardjohn) [reply manifests/charts/istio-cni/templates/daemonset.yaml:173]: Well not really actually, we just moved the problem into the code with `s.cleanupNode` (I think)

**Review by howardjohn — APPROVED** [review]

File: manifests/charts/istio-cni/templates/daemonset.yaml (hunk 1 above)
[the remark this answers is not in the document]
Comment (bleggett) [reply manifests/charts/istio-cni/templates/daemonset.yaml:173]: Correct - I think the problem is still (probably) the same - the logic is the same in both spots, we were just doing the same thing in two spots before, which seemed unnecessary.

 This also might be an aspect of the "how we handle ztunnel upgrades with CNI" problem.

[the remark this answers is not in the document]
Comment (howardjohn) [reply manifests/charts/istio-cni/templates/daemonset.yaml:173]: Actually thinking about it more.. the go code was "ztunnel shutdown". This is "cni shutdown".

If the CNI gets killed with `remove --force` or whatever the node may be permanently stuck. preStop probably doesn't actually prevent that, either, though.

[the remark this answers is not in the document]
Comment (bleggett) [reply manifests/charts/istio-cni/templates/daemonset.yaml:173]: Yeah neither would be foolproof. This aspect needs a better solution generally either way. We would be better off "clean slating" as an initialization precheck than relying solely on cleanup hooks to work reliably IMO.

[the remark this answers is not in the document]
Comment (howardjohn) [reply manifests/charts/istio-cni/templates/daemonset.yaml:173]: I think the issue is if they totally uninstall Istio but yeah I agree. There is no way to _guarantee_ we run cleanup logic on shutdown anyways

[the remark this answers is not in the document]
Comment (bleggett) [reply manifests/charts/istio-cni/templates/daemonset.yaml:173]: Lemme try that real quick - I can add this back if we want it as a last-ditch uninstall hook.

[the remark this answers is not in the document]
Comment (howardjohn) [reply manifests/charts/istio-cni/templates/daemonset.yaml:173]: I was worried about this preStop hook specifically for a rolling restart of CNI (upgrade) causing an outage

**Pull request merged**